File: polanyi/scripts/python/revenue_bridge.py
#!/usr/bin/env python3
"""
Revenue Bridge
Translates media metrics (L3) to business metrics (L2) to financial metrics (L1).
Generates CEO/CFO/CMO summaries and efficiency scores.

Usage: echo '{"command":"translate","data":{...}}' | python3 revenue_bridge.py
Commands: translate, compare
"""

import json
import sys


def translate(data):
    """Translate L3 media metrics through L2 to L1 financial metrics."""
    channels = data.get("channels", [])
    aov = data.get("aov", 100)
    retention_rate = data.get("retention_rate", 0.3)
    avg_customer_months = data.get("avg_customer_months", 12)

    total_spend = 0
    total_revenue = 0
    total_conversions = 0
    total_impressions = 0
    total_clicks = 0
    channel_results = []

    for ch in channels:
        spend = ch.get("spend", 0)
        impressions = ch.get("impressions", 0)
        clicks = ch.get("clicks", 0)
        conversions = ch.get("conversions", 0)
        revenue = ch.get("revenue", conversions * aov)

        total_spend += spend
        total_revenue += revenue
        total_conversions += conversions
        total_impressions += impressions
        total_clicks += clicks

        l3 = {}
        if impressions > 0:
            l3["cpm"] = round((spend / impressions) * 1000, 2)
        if clicks > 0:
            l3["cpc"] = round(spend / clicks, 2)
        if impressions > 0:
            l3["ctr"] = round((clicks / impressions) * 100, 2)
        if clicks > 0:
            l3["cvr"] = round((conversions / clicks) * 100, 2)

        l2 = {}
        if conversions > 0:
            l2["cpa"] = round(spend / conversions, 2)
        if spend > 0:
            l2["roas"] = round(revenue / spend, 2)
        l2["aov"] = round(revenue / conversions, 2) if conversions > 0 else 0
        l2["conversions"] = conversions

        channel_results.append({
            "channel": ch["name"],
            "spend": spend,
            "revenue": revenue,
            "l3_metrics": l3,
            "l2_metrics": l2,
        })

    # L1 Financial Metrics
    cac = round(total_spend / total_conversions, 2) if total_conversions > 0 else 0
    clv = round(aov * retention_rate * avg_customer_months, 2)
    clv_cac_ratio = round(clv / cac, 2) if cac > 0 else 0
    roi = round(((total_revenue - total_spend) / total_spend) * 100, 1) if total_spend > 0 else 0
    marketing_margin = round(((total_revenue - total_spend) / total_revenue) * 100, 1) if total_revenue > 0 else 0

    l1 = {
        "cac": cac,
        "clv": clv,
        "clv_cac_ratio": clv_cac_ratio,
        "roi_pct": roi,
        "marketing_margin_pct": marketing_margin,
        "total_revenue": total_revenue,
        "total_spend": total_spend,
    }

    # Efficiency Score (0-100)
    scores = []
    if clv_cac_ratio >= 3:
        scores.append(100)
    elif clv_cac_ratio >= 1:
        scores.append(round(clv_cac_ratio / 3 * 100))
    else:
        scores.append(0)

    roas = total_revenue / total_spend if total_spend > 0 else 0
    if roas >= 4:
        scores.append(100)
    elif roas >= 1:
        scores.append(round(roas / 4 * 100))
    else:
        scores.append(0)

    if marketing_margin > 0:
        scores.append(min(100, round(marketing_margin * 2)))
    else:
        scores.append(0)

    efficiency_score = round(sum(scores) / len(scores)) if scores else 0

    # C-Suite Summaries
    csuite = {
        "ceo": (
            f"Marketing investment of ${total_spend:,.0f} generated "
            f"${total_revenue:,.0f} in revenue ({total_conversions:,} customers). "
            f"Customer lifetime value is ${clv:,.0f} vs acquisition cost of ${cac:,.0f} "
            f"(CLV:CAC = {clv_cac_ratio}:1). "
            f"ROI: {roi}%."
        ),
        "cfo": (
            f"Total spend: ${total_spend:,.0f}. Revenue: ${total_revenue:,.0f}. "
            f"Marketing margin: {marketing_margin}%. ROI: {roi}%. "
            f"Blended ROAS: {roas:.2f}x. "
            f"Efficiency score: {efficiency_score}/100."
        ),
        "cmo": (
            f"Campaign delivered {total_conversions:,} conversions across "
            f"{len(channels)} channels at ${cac:,.0f} CPA. "
            f"ROAS: {roas:.2f}x. "
            f"Top channel by ROAS: "
            f"{max(channel_results, key=lambda x: x['l2_metrics'].get('roas', 0))['channel'] if channel_results else 'N/A'}."
        ),
    }

    return {
        "l1_metrics": l1,
        "efficiency_score": efficiency_score,
        "channels": channel_results,
        "csuite_summary": csuite,
    }


def compare(data):
    """Compare channels side by side."""
    channels = data.get("channels", [])
    results = []
    for ch in channels:
        spend = ch.get("spend", 0)
        revenue = ch.get("revenue", 0)
        conversions = ch.get("conversions", 0)
        results.append({
            "channel": ch["name"],
            "spend": spend,
            "revenue": revenue,
            "roas": round(revenue / spend, 2) if spend > 0 else 0,
            "cpa": round(spend / conversions, 2) if conversions > 0 else 0,
            "efficiency_rank": 0,
        })
    results.sort(key=lambda x: x["roas"], reverse=True)
    for i, r in enumerate(results):
        r["efficiency_rank"] = i + 1
    return {"comparison": results}


def main():
    if len(sys.argv) > 1 and sys.argv[1] == "--help":
        print(json.dumps({
            "usage": "echo '{\"command\":\"translate\",\"data\":{...}}' | python3 revenue_bridge.py",
            "commands": ["translate", "compare"],
        }, indent=2))
        return

    raw = sys.stdin.read().strip()
    if not raw:
        print(json.dumps({"error": "No input. Use --help for usage."}))
        sys.exit(1)

    payload = json.loads(raw)
    cmd = payload.get("command", "")
    data = payload.get("data", {})

    if cmd == "translate":
        result = translate(data)
    elif cmd == "compare":
        result = compare(data)
    else:
        result = {"error": f"Unknown command: {cmd}. Valid: translate, compare"}

    print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()

File: openagency/packages/engines/src/executive-bridge/revenue-bridge.ts
// ─── Revenue Bridge ─────────────────────────────────────────────────
// Translates L3 media -> L2 business -> L1 financial metrics.
// L3→L2→L1 revenue translation with CLV computation

import { round } from '@openagency/core/utils/math';
import type {
  RevenueBridgeInput,
  RevenueBridgeOutput,
  L1Metrics,
  L2Metrics,
  L3Metrics,
  CSuiteSummary,
} from '@openagency/types';

export function translate(data: RevenueBridgeInput): RevenueBridgeOutput {
  const channels = data.channels ?? [];
  const aov = data.aov ?? 100;
  const retentionRate = data.retention_rate ?? 0.3;
  const avgCustomerMonths = data.avg_customer_months ?? 12;
  const baselinePct = data.baseline_pct; // 0-1 fraction, e.g. 0.535 = 53.5% organic

  let totalSpend = 0;
  let totalRevenue = 0;
  let totalConversions = 0;

  const channelResults: RevenueBridgeOutput['channels'] = [];

  for (const ch of channels) {
    const spend = ch.spend ?? 0;
    const impressions = ch.impressions ?? 0;
    const clicks = ch.clicks ?? 0;
    const conversions = ch.conversions ?? 0;
    const revenue = ch.revenue ?? conversions * aov;

    totalSpend += spend;
    totalRevenue += revenue;
    totalConversions += conversions;

    const l3: L3Metrics = {};
    if (impressions > 0) l3.cpm = round((spend / impressions) * 1000);
    if (clicks > 0) l3.cpc = round(spend / clicks);
    if (impressions > 0) l3.ctr = round((clicks / impressions) * 100);
    if (clicks > 0) l3.cvr = round((conversions / clicks) * 100);

    const l2: L2Metrics = {
      cpa: conversions > 0 ? round(spend / conversions) : 0,
      roas: spend > 0 ? round(revenue / spend) : 0,
      aov: conversions > 0 ? round(revenue / conversions) : 0,
      conversions,
    };

    channelResults.push({
      channel: ch.name,
      spend,
      revenue,
      l3_metrics: l3,
      l2_metrics: l2,
    });
  }

  // L1 Financial Metrics (raw — assumes 100% media-driven)
  const cac = totalConversions > 0 ? round(totalSpend / totalConversions) : 0;
  const clv = round(aov * retentionRate * avgCustomerMonths);
  const clvCacRatio = cac > 0 ? round(clv / cac) : 0;
  const roi = totalSpend > 0 ? round(((totalRevenue - totalSpend) / totalSpend) * 100, 1) : 0;
  const marketingMargin = totalRevenue > 0
    ? round(((totalRevenue - totalSpend) / totalRevenue) * 100, 1)
    : 0;
  const roas = totalSpend > 0 ? totalRevenue / totalSpend : 0;

  const l1: L1Metrics = {
    cac,
    clv,
    clv_cac_ratio: clvCacRatio,
    roi_pct: roi,
    marketing_margin_pct: marketingMargin,
    total_revenue: totalRevenue,
    total_spend: totalSpend,
  };

  // Incremental metrics (MMM-adjusted) when baseline_pct is provided
  if (baselinePct != null && baselinePct >= 0 && baselinePct < 1) {
    const incrementalRevenue = round(totalRevenue * (1 - baselinePct));
    const iroas = totalSpend > 0 ? round(incrementalRevenue / totalSpend) : 0;
    const iroi = totalSpend > 0 ? round(((incrementalRevenue - totalSpend) / totalSpend) * 100, 1) : 0;
    const imargin = incrementalRevenue > 0
      ? round(((incrementalRevenue - totalSpend) / incrementalRevenue) * 100, 1)
      : 0;
    // Incremental conversions proportional to incremental revenue share
    const iConversions = totalRevenue > 0
      ? totalConversions * (incrementalRevenue / totalRevenue)
      : 0;
    const icac = iConversions > 0 ? round(totalSpend / iConversions) : 0;
    const iclvCac = icac > 0 ? round(clv / icac) : 0;

    l1.incremental = {
      baseline_pct: baselinePct,
      incremental_revenue: incrementalRevenue,
      iroas,
      iroi_pct: iroi,
      imarketing_margin_pct: imargin,
      iclv_cac_ratio: iclvCac,
      icac,
    };
  }

  // Efficiency Score — use incremental metrics when available, else raw
  const effectiveRoas = l1.incremental ? l1.incremental.iroas : roas;
  const effectiveClvCac = l1.incremental ? l1.incremental.iclv_cac_ratio : clvCacRatio;
  const effectiveMargin = l1.incremental ? l1.incremental.imarketing_margin_pct : marketingMargin;

  const scores: number[] = [];
  if (effectiveClvCac >= 3) scores.push(100);
  else if (effectiveClvCac >= 1) scores.push(Math.round((effectiveClvCac / 3) * 100));
  else scores.push(0);

  if (effectiveRoas >= 4) scores.push(100);
  else if (effectiveRoas >= 1) scores.push(Math.round((effectiveRoas / 4) * 100));
  else scores.push(0);

  if (effectiveMargin > 0) scores.push(Math.min(100, Math.round(effectiveMargin * 2)));
  else scores.push(0);

  const efficiencyScore = scores.length > 0
    ? Math.round(scores.reduce((a, b) => a + b, 0) / scores.length)
    : 0;

  // C-Suite Summaries — use incremental when available
  const topChannel = channelResults.length > 0
    ? channelResults.reduce((best, ch) =>
        (ch.l2_metrics.roas > best.l2_metrics.roas ? ch : best))
    : null;

  const inc = l1.incremental;
  const csuite: CSuiteSummary = inc
    ? {
        ceo: `Marketing investment of $${totalSpend.toLocaleString()} generated $${totalRevenue.toLocaleString()} total revenue, of which $${inc.incremental_revenue.toLocaleString()} is incremental (${round((1 - inc.baseline_pct) * 100, 1)}% media-driven, ${round(inc.baseline_pct * 100, 1)}% organic baseline). iROI: ${inc.iroi_pct}%. iCLV:CAC = ${inc.iclv_cac_ratio}:1.`,
        cfo: `Total spend: $${totalSpend.toLocaleString()}. Incremental revenue: $${inc.incremental_revenue.toLocaleString()} (baseline ${round(inc.baseline_pct * 100, 1)}% excluded). iROAS: ${inc.iroas.toFixed(2)}x. iMarketing margin: ${inc.imarketing_margin_pct}%. Efficiency score: ${efficiencyScore}/100.`,
        cmo: `Campaign delivered ${totalConversions.toLocaleString()} total conversions across ${channels.length} channels. iROAS: ${inc.iroas.toFixed(2)}x (vs raw ${roas.toFixed(2)}x). Top channel: ${topChannel?.channel ?? 'N/A'}. Organic baseline: ${round(inc.baseline_pct * 100, 1)}%.`,
      }
    : {
        ceo: `Marketing investment of $${totalSpend.toLocaleString()} generated $${totalRevenue.toLocaleString()} in revenue (${totalConversions.toLocaleString()} customers). CLV: $${clv.toLocaleString()} vs CAC: $${cac.toLocaleString()} (CLV:CAC = ${clvCacRatio}:1). ROI: ${roi}%. ⚠️ No baseline deduction — assumes 100% media-driven.`,
        cfo: `Total spend: $${totalSpend.toLocaleString()}. Revenue: $${totalRevenue.toLocaleString()}. Marketing margin: ${marketingMargin}%. ROI: ${roi}%. ROAS: ${roas.toFixed(2)}x. ⚠️ Pass baseline_pct from MMM for incremental metrics.`,
        cmo: `Campaign delivered ${totalConversions.toLocaleString()} conversions across ${channels.length} channels at $${cac.toLocaleString()} CPA. ROAS: ${roas.toFixed(2)}x. Top channel: ${topChannel?.channel ?? 'N/A'}. ⚠️ Raw metrics — no organic baseline applied.`,
      };

  const result: RevenueBridgeOutput = {
    l1_metrics: l1,
    efficiency_score: efficiencyScore,
    channels: channelResults,
    csuite_summary: csuite,
  };

  if (!inc) {
    result.baseline_warning =
      'Metrics assume 100% of revenue is media-driven. Pass baseline_pct (0-1) from MMM baseline decomposition for incremental metrics (iROAS, iROI, iCLV:CAC). Without baseline deduction, ROAS/ROI are inflated.';
  }

  return result;
}

export function compareChannels(data: {
  channels: Array<{ name: string; spend: number; revenue: number; conversions: number }>;
}) {
  const results = data.channels.map((ch) => ({
    channel: ch.name,
    spend: ch.spend,
    revenue: ch.revenue,
    roas: ch.spend > 0 ? round(ch.revenue / ch.spend) : 0,
    cpa: ch.conversions > 0 ? round(ch.spend / ch.conversions) : 0,
    efficiency_rank: 0,
  }));
  results.sort((a, b) => b.roas - a.roas);
  results.forEach((r, i) => (r.efficiency_rank = i + 1));
  return { comparison: results };
}
